src/infill/SierpinskiFill.cpp

```diff
@@ -0,0 +1,802 @@
+    float total_remaining_value_error = 0;
+    for (unsigned int remaining_node_order_idx = node_order_idx; remaining_node_order_idx < nodes.size(); remaining_node_order_idx++)
+    {
+        int node_idx = order[remaining_node_order_idx];
+        SierpinskiTriangle* node = nodes[node_idx];
+        float val_err = node->getValueError();
+        assert(val_err > -allowed_length_error);
+        total_remaining_value_error += val_err;
+    }
+    
+    if (total_remaining_value_error < added - allowed_length_error)
+    {
+        std::cerr << "total_remaining:" << total_remaining_value_error << " should be > " << added << '\n';
+        assert(false);
+    }
+    
+    if (std::abs(total_remaining_value_error) < .0001)
+    {
+        return;
+    }
+
+    for (unsigned int remaining_node_order_idx = node_order_idx; remaining_node_order_idx < nodes.size(); remaining_node_order_idx++)
+    {
+        int node_idx = order[remaining_node_order_idx];
+        SierpinskiTriangle* node = nodes[node_idx];
+        float val_error = node->getValueError();
+        assert(val_error > -allowed_length_error);
+        
+        float diff = added * val_error / total_remaining_value_error;
+        subtracted += diff;
+        node_error_compensation[node_idx] = -diff;
+    }
+    if (std::abs(subtracted - added) > allowed_length_error)
+    {
+        std::cerr << "Redistribution didn't distribute well!! added " << added << " subtracted" << subtracted << '\n';
+        assert(false);
+    }
+
+    float energy = 0;
+    for (unsigned int node_idx = 0; node_idx < nodes.size(); node_idx++)
+    {
+        nodes[node_idx]->error_left -= energy;
+        energy += node_error_compensation[node_idx];
+        nodes[node_idx]->error_right += energy;
+    }
+    assert(energy < allowed_length_error);
+}
+
+void SierpinskiFill::settleErrors()
+{
+    std::vector<std::list<std::list<SierpinskiTriangle*>::iterator>> depth_ordered;
+    { // compute depth_ordered
+        for (int i = 0; i < max_depth + 1; i++)
+        {
+            depth_ordered.emplace_back();
+        }
+        for (std::list<SierpinskiTriangle*>::iterator it = sequence.begin(); it != sequence.end(); ++it)
+        {
+            SierpinskiTriangle* node = *it;
+            depth_ordered[node->depth].emplace_back(it);
+        }
+    }
+
+    for (std::list<std::list<SierpinskiTriangle*>::iterator>& depth_nodes : depth_ordered)
+        for (std::list<SierpinskiTriangle*>::iterator it : depth_nodes)
+        {
+            redistributeLeftoverErrors(it, std::next(it), false);
+        }
+}
+
+void SierpinskiFill::diffuseError()
+{
+    int pair_constrained_nodes = 0;
+    int constrained_nodes = 0;
+    int unconstrained_nodes = 0;
+    int subdivided_nodes = 0;
+    float error = 0;
+    for (std::list<SierpinskiTriangle*>::iterator it = sequence.begin(); it != sequence.end(); ++it)
+    {
+        SierpinskiTriangle* node = *it;
+        SierpinskiTriangle& triangle = *node;
```

Comment:
There's no difference between these two. Sometimes you're using `node->` and sometimes you're using `triangle.`. I suggest you use only `triangle.` (`triangle = **it`) or only `node->`.

Reply:
Ok: 069d19ca783728a2aefb170fc7553e31cf4a972a